File: workflow/scripts/ProgressiveGenes.R
#!/usr/bin/env Rscript
log <- file(snakemake@log[[1]], open="wt")
sink(log)
sink(log, type="message")


#' Finds genes that are differentially expressed in the same direction
#' Across two DE comparisons. For example, a gene may be significantly 
#' upregulated in the control v intermediate phenotype comparison. 
#' And also in the intermediate phenotype vs full phenotype comparison.

library(data.table)
library(tidyverse)
library(glue)

# read metadata and get contrasts
comps = snakemake@params['comps'][[1]]
padj_threshold = snakemake@params['pval']
isoform_level = snakemake@params['isoform_level']
gene_level = snakemake@params['gene_level']
upper_fc = snakemake@params['fc']
lower_fc = 1/as.numeric(upper_fc) # if someone wants a FC threshold of 2, need to have lower threshold of 0.5.

for (cont in comps){
  sus = str_split(cont, "_")[[1]][1]           # get control/susceptible name (such as Kisumu)
  intermediate = str_split(cont, "_")[[1]][2]     # get intermediate
  res = str_split(cont, "_")[[1]][3]              # get last of string, which is resistant/case 

  if (gene_level == TRUE){

    #### Gene diff ####
    one = fread(glue("results/genediff/{intermediate}_{res}.csv"))
    up1  = one %>% filter(FC > upper_fc, padj < padj_threshold)
    down1 = one %>% filter(FC < lower_fc, padj < padj_threshold)
    
    two = fread(glue("results/genediff/{sus}_{intermediate}.csv"))
    up2  = two %>% filter(FC > upper_fc, padj < padj_threshold)
    down2 = two %>% filter(FC < lower_fc, padj < padj_threshold)
    
    intersectdown = inner_join(down1, down2, by="GeneID", suffix=c("_field", "_lab"))
    intersectup = inner_join(up1, up2, by="GeneID", suffix=c("_field", "_lab"))
    
    fwrite(intersectup, glue("results/genediff/{sus}_{intermediate}_{res}.up.progressive.tsv"), sep="\t")
    fwrite(intersectdown, glue("results/genediff/{sus}_{intermediate}_{res}.down.progressive.tsv"), sep="\t")
  }

  if (isoform_level == TRUE){
    #### Isoforms #### 
    one = fread(glue("results/isoformdiff/{intermediate}_{res}.csv"))
    up1  = one %>% filter(FC > upper_fc, qval < padj_threshold)
    down1 = one %>% filter(FC < lower_fc, qval < padj_threshold)
    
    two = fread(glue("results/isoformdiff/{sus}_{intermediate}.csv"))
    up2  = two %>% filter(FC > upper_fc, qval < padj_threshold)
    down2 = two %>% filter(FC < lower_fc, qval < padj_threshold)
    
    intersectdown = inner_join(down1, down2, by="TranscriptID", suffix=c("_field", "_lab"))
    intersectup = inner_join(up1, up2, by="TranscriptID", suffix=c("_field", "_lab"))
    
    fwrite(intersectup, glue("results/isoformdiff/{sus}_{intermediate}_{res}.up.progressive.tsv"), sep="\t")
    fwrite(intersectdown, glue("results/isoformdiff/{sus}_{intermediate}_{res}.down.progressive.tsv"), sep="\t")
  }

}


sessionInfo()

File: workflow/scripts/misc/progressive-genes.py
#!/usr/bin/env python3

import argparse
import pandas as pd
from pathlib import Path

def parse_args():
    parser = argparse.ArgumentParser(
        description='Find genes/isoforms that are differentially expressed in the same direction across comparisons'
    )
    parser.add_argument('--comps', nargs='+', required=True,
                       help='Comparisons in format control_intermediate_resistant')
    parser.add_argument('--padj-threshold', type=float, default=0.05,
                       help='Adjusted p-value threshold')
    parser.add_argument('--fc-threshold', type=float, default=2.0,
                       help='Fold change threshold')
    parser.add_argument('--gene-level', action='store_true',
                       help='Analyze gene-level differential expression')
    parser.add_argument('--isoform-level', action='store_true',
                       help='Analyze isoform-level differential expression')
    return parser.parse_args()

def process_comparison(sus, intermediate, res, padj_threshold, upper_fc, lower_fc, level='gene'):
    """Process a single comparison for either genes or isoforms."""
    
    # Set paths and column names based on analysis level
    if level == 'gene':
        base_path = 'results/genediff'
        id_col = 'GeneID'
        padj_col = 'padj'
    else:  # isoform
        base_path = 'results/isoformdiff'
        id_col = 'TranscriptID'
        padj_col = 'qval'
    
    # Read the comparison files
    one = pd.read_csv(f"{base_path}/{intermediate}_{res}.csv")
    two = pd.read_csv(f"{base_path}/{sus}_{intermediate}.csv")
    
    # Filter for up and down regulated genes/isoforms
    up1 = one[
        (one['FC'] > upper_fc) & 
        (one[padj_col] < padj_threshold)
    ]
    down1 = one[
        (one['FC'] < lower_fc) & 
        (one[padj_col] < padj_threshold)
    ]
    
    up2 = two[
        (two['FC'] > upper_fc) & 
        (two[padj_col] < padj_threshold)
    ]
    down2 = two[
        (two['FC'] < lower_fc) & 
        (two[padj_col] < padj_threshold)
    ]
    
    # Find intersections
    intersect_up = pd.merge(
        up1, up2,
        on=id_col,
        suffixes=('_field', '_lab')
    )
    intersect_down = pd.merge(
        down1, down2,
        on=id_col,
        suffixes=('_field', '_lab')
    )
    
    # Save results
    output_base = f"{base_path}/{sus}_{intermediate}_{res}"
    intersect_up.to_csv(f"{output_base}.up.progressive.tsv", sep='\t', index=False)
    intersect_down.to_csv(f"{output_base}.down.progressive.tsv", sep='\t', index=False)

def main():
    args = parse_args()
    
    # Calculate lower fold change threshold
    upper_fc = args.fc_threshold
    lower_fc = 1.0 / upper_fc
    
    # Process each comparison
    for comp in args.comps:
        sus, intermediate, res = comp.split('_')
        
        if args.gene_level:
            process_comparison(
                sus, intermediate, res,
                args.padj_threshold, upper_fc, lower_fc,
                level='gene'
            )
            
        if args.isoform_level:
            process_comparison(
                sus, intermediate, res,
                args.padj_threshold, upper_fc, lower_fc,
                level='isoform'
            )

if __name__ == '__main__':
    main()